### PR TITLE
Catch Panic and Error from Atom.to_text and yield detailed error message

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Don't propagate errors from `toDisplayString` - construct an error message with `Atom.toString`. 

### Important Notes

> currently a failure in to_text is swallowed by `toString` and we cannot detect that something went wrong during the serialization

Not sure how satisfying the solution is, but the error swallowing happens in Truffle and there is little to do with it. We can just catch the error ourselves and produce some meaningful string.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/List.enso [hunk 1]
```diff
@@ -404,10 +404,10 @@ type List
 
     to_text : Text
     to_text self =
-        go l t = case l of
-            Nil -> t + "Nil"
-            Cons x xs -> @Tail_Call go xs (t + "Cons " + x.to_text + " ")
-        go self ""
+        go l t e = case l of
```
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/List.enso:407]: This is unrelated to [pivotal 183578531](https://www.pivotaltracker.com/n/projects/2539304/stories/182522644) but fixes a problem introduced by #4023  I realized during Monday's demo meeting.

File: engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+        to_text : Text
+        to_text self = Panic.throw (Illegal_State.Error "C")
+
+    panic x = case x of
+        1 -> panic1
+        2 -> panic2
+        _ -> panic3
+
+    panic1 = PolyglotErrorTest.bar (TypeCa.Ca 'x')
```
Comment (JaroslavTulach) [inline engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java:83]: @radeusgd, if you want me to cover more cases, feel free to expand the test and I do my best to deal with the additional situations.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java [hunk 3]
```diff
@@ -183,6 +184,7 @@ Text toDisplayString(boolean allowSideEffects, @CachedLibrary("this") InteropLib
     try {
       return TypesGen.expectText(atoms.invokeMember(this, "to_text"));
     } catch (UnsupportedMessageException
+        | AbstractTruffleException
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: **Update**: Since https://github.com/enso-org/enso/commit/5fa1b5c98d2615ad567ad6a3c4da44bd129f4229 we recognize the Panic/Error and construct verbose message.

This is a relatively easy fix. There already was a previous behavior that on (almost) any exception fell back to `atom.toString(10)`. It even worked for Enso `Error` as it was caught as `UnexpectedResultException`. I am just expanding it to `Panic` and other guest languages exceptions.

File: engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+import static org.junit.Assert.assertTrue;
+import org.junit.Before;
+import org.junit.Test;
+
+public class PolyglotErrorTest {
+  private Context ctx;
+  private Value panic;
+
+  public static String bar(Object o) {
+    return "[[" + o + "]]";
```
Comment (JaroslavTulach) [inline engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java:22]: If one wants to control the exception behavior then I am afraid one has to use `Value` - for example:
```diff
--- a/engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java
+++ b/engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java
@@ -18,9 +18,22 @@
   private Context ctx;
   private Value panic;
 
-  public static String bar(Object o) {
-    return "[[" + o + "]]";
+  public static String bar(Value v) throws Exception {
+    String s;
+    if (v.isString()) {
+      s = v.asString();
+    } else {
+      var t = v.invokeMember("to_text");
+      if (t.isString()) {
+        s = t.asString();
+      } else if (t.isException()) {
+        throw t.as(Exception.class);
+      } else {
+        s = t.toString();
   }
+    }
+    return "[[" + s + "]]";
+  }
 
   @Before
   public void prepareCtx() throws Exception {
```
allows one to observe `Panic` (via `catch PolyglotException`) and `Error` (via `isException()`).

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java (hunk 3 above)
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Ok, so we catch the error here - so we _do_ have point where we see that `to_text` failed, right? Cannot we then throw a Panic exception from within here instead of returning the fallback error?

File: engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java (hunk 2 above)
Comment (radeusgd) [reply engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java:83]: The test cases are ok, but I'm not sure if the behavior is what we want.

Of course, if there is no way to throw a panic in case `to_text` failed with a panic/error, then so be it. But looking at the code above, can we throw instead of returning the fallback? If not, could you please explain to me why not?

I'd value @jdunkerley's input here as well. I think in some cases a fallback text representation is actually useful (for example when I'm trying to debug something, if I get an error in `to_text` I won't even immediately know where it happened without some fallback). On the other hand, from the libraries perspective I think it is very important to be able to detect these errors.

Consider case the user saves some custom object in a Table, whose `to_text` is broken and is erroring. The user is serializing such a table to a CSV file. I think I'd rather get an error/warning saying that some objects failed to serialize (and e.g. in warning mode I'd replace the problematic entries with `null`/`""` instead of some fallback which may not always be appropriate).

Ideally, I think it would be best if we could throw some special exception `Unexpected_Failure_In_To_Text` which would contain within its message/payload that mentioned fallback text representation - this way we will kill 2 birds with one stone - we'll still be able to handle the errors in our libraries, but for example when debugging we will have at least some concept of what is the object whose `to_text` was broken.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java (hunk 3 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: We can throw an exception, but it ends up at [truffle catch block](https://github.com/oracle/graal/blob/83c37dde9922be071b8d6efe3b0a140bbb1231f2/truffle/src/com.oracle.truffle.polyglot/src/com/oracle/truffle/polyglot/PolyglotWrapper.java#L181) - there is not much to do about it.

Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Ah, I see! Well that's too bad. In that case, I guess the current solution is right.

Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: However, I feel like it could be good to at least signalize that the default `to_text` method has failed to execute and we are getting a replacement instead - personally I'm very much against returning such replacements without telling the user.

Could we maybe change the replacement to to something like
```
[Error in method to_text of <fallback-toString>: <Exception message>]
```
?

I think this could be the best middle ground - we still return some valid text representation, we still show the fallback representation so that the user can know what was the object that failed - and when we include the exception message too, the user can also have some clue as to what has happened (of course it's not as good as with a stacktrace, but including the stack trace would make the representation too long, so I think just the message is a reasonable middle ground).

Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: So I'd suggest to amend the 2nd and 3rd tests to be:

```
assertEquals("[[[Error in method `to_text` of [Cb y]: Illegal_State.Error \"B\"]]]", v.asString());
```
and
```
assertEquals("[[[Panic in method `to_text` of [Cc z]: Illegal_State.Error \"C\"]]]", v.asString());
```
respectively.

Ideally we should distinguish a dataflow error (writing `Error in method` or even `Dataflow error in method`) and panics (`Panic in method`). We could also use `Exception in method` if it was a Java exception, but I guess since Java exceptions are the same as Panics for Enso, I think just Panic is OK.

What do you think? Does it sound OK?

File: engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java (hunk 2 above)
Comment (JaroslavTulach) [reply engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java:83]: Throwing `Unexpected_Failure_In_To_Text` isn't possible in version 22.3. I've just started [discussion on GraalVM slack channel](https://graalvm.slack.com/archives/CNQSB2DHD/p1673413554022339?thread_ts=1673412770.137559&cid=CNQSB2DHD) to bring this issue to GraalVM guys attention and make it possible in future versions of GraalVM.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java (hunk 3 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: 5fa1b5c98d2615ad567ad6a3c4da44bd129f4229 is the closest I could get.

**Review by hubertp — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java [hunk 5]
```diff
@@ -179,16 +193,25 @@ static Object doUncached(
...
+      handleError.enter();
+      msg = this.toString("Panic in method `to_text` of [", 10, "]: ", panic);
+    } catch (UnexpectedResultException dataflowError) {
+      handleError.enter();
+      msg = this.toString("Error in method `to_text` of [", 10, "]: ", result);
+    } catch (
+      UnsupportedMessageException | ArityException | UnknownIdentifierException | UnsupportedTypeException e
+    ) {
+      handleError.enter();
+      msg = this.toString(null, 10, null, null);
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Why not do the `Panic in method` for these exceptions?

File: engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java (hunk 2 above)
Comment (radeusgd) [reply engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java:83]: Thanks for bringing up the topic on Slack.

The changes to the implementation made me realise it would be good to at least handle one more case:

```
type TypeCd
    Cd x

    to_text self = 42

panic4 = PolyglotErrorTest.bar (TypeCd.Cd Nothing)
```

I don't have a very particular requirement of what the message should be, but it should clearly indicate to the user that an unexpected type has been provided.

I'd suggest something along the lines of:
```
[Error in method `to_text` of [TypeCd.Cd Nothing]: Type_Error: expected ??? to be Text but got 42.]
```

We probably need to amend the `UnexpectedResultException` branch and change the error message depending on if `result` is a dataflow error or if it contains any other value.

To make sure the logic is correctly implemented ideally we'd also have these tests:

```
type TypeCe
    Cd x

    to_text self =
        Warning.attach "Some random warning" self.x

text5 = PolyglotErrorTest.bar (TypeCd.Ce "Foo")
panic6 = PolyglotErrorTest.bar (TypeCd.Ce 44)
```

I'd then expect `text5` to just be equal to `Foo` - i.e. I want to check that attaching a warning within the `to_text` call does not break the interop `toString`.

And `panic6` should return a result consistent with `panic4` - i.e. it should say that it expected a Text but got a `44`. It should not display something like `WithWarnings@deadbeef`.

Comment (JaroslavTulach) [reply engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java:83]: Tests up to `panic6` implemented in [4997b24](https://github.com/enso-org/enso/pull/4034/commits/4997b246a92e709ad0290c6ca6dadca97603a00c).

Comment (radeusgd) [reply engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java:83]: Thanks!

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java [hunk 6]
```diff
@@ -179,16 +195,34 @@ static Object doUncached(
...
+        return TypesGen.expectText(result);
+      }
+    } catch (AbstractTruffleException panic) {
+      handleError.enter();
+      msg = this.toString("Panic in method `to_text` of [", 10, "]: ", panic);
+    } catch (
+      UnexpectedResultException | UnsupportedMessageException | ArityException | UnknownIdentifierException | UnsupportedTypeException e
+    ) {
+      handleError.enter();
+      msg = this.toString("Error in method `to_text` of [", 10, "]: Expected Text but got ", result);
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Sorry to be so picky, I think it's the last one,

I'm a bit surprised why other kinds of exceptions are handled the same way as `UnexpectedResultException` - that one is about a wrong type so `Expected Text but got ...` is just right for it, but it is a bit weird for the other errors, no?

I'd keep `Panic in method` or `Exception in method` for the others, without the `Expected Text`, and keep the `Expected Text` just for type errors (`UnexpectedResultException`).

Sorry that I cannot provide a decent test case for these other exceptions, I'm not sure how to trigger them in _that_ context.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java [hunk 7]
```diff
@@ -86,13 +90,25 @@ private void toString(StringBuilder builder, boolean shouldParen, int depth) {
...
+    }
     toString(sb, false, depth);
+    if (suffix != null) {
+      sb.append(suffix);
+    }
+    if (obj != null) {
+      var errorMessage = InteropLibrary.getUncached().toDisplayString(obj);
+      if (errorMessage != null) {
+        sb.append(errorMessage);
+      }
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]:
```suggestion
      if (errorMessage != null) {
        sb.append(errorMessage);
      } else {
        sb.append("Unknown error.");
      }
```

As our messages all say stuff along the lines of `Error in method `to_text` of <foo>:` -so we have this colon suggesting something is to follow, I think in the `null` case we still should have something after that colon, so that the message does not look cut off halfway through.

**Review by radeusgd — APPROVED** [review]
While we are still discussing if we want the fallbacks or propagation of the errors, even if we file the issue to the Graal team to implement the error propagation, it will take some time until it's implemented and until we switch to a newer Graal version having it. So these fallbacks are definitely a great improvement from the previous state - whether temporary or to stay for longer, I think it will be helpful to get them integrated.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java (hunk 6 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: > I cannot provide a decent test case for these other exceptions, I'm not sure how to trigger them in that context.

That's my problem as well. Triggering these exceptions isn't easy, but they have to be caught, as they are checked exception.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java (hunk 7 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Printing `Nothing` in [4e55305](https://github.com/enso-org/enso/pull/4034/commits/4e55305e688d69a842ebffde063694bba22c4fcf)

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java (hunk 6 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Rewritten in [44978cd](https://github.com/enso-org/enso/pull/4034/commits/44978cd0c59705bc662a75a1fd14db4cd83c9398). Now it will print `Panic...` in case of those _strange exceptions_.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java (hunk 5 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: It is panic since [44978cd](https://github.com/enso-org/enso/pull/4034/commits/44978cd0c59705bc662a75a1fd14db4cd83c9398)

File: engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java (hunk 4 above)
Comment (JaroslavTulach) [reply engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java:22]: @radeusgd, regardless of the [outcome of the GraalVM slack discussion](https://graalvm.slack.com/archives/CNQSB2DHD/p1673436829788349?thread_ts=1673412770.137559&cid=CNQSB2DHD), the previous snippet gives you a way to control, propagate and handle exceptions from `to_text` if necessary. Works in current GraalVM version.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java (hunk 6 above)
Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Perfect, thanks!

File: engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java (hunk 4 above)
Comment (radeusgd) [reply engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java:22]: Oh, that is a pretty sensible workaround! Thanks for showing it and also I somehow missed it.

File: engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+      .allowAllAccess(true)
+      .logHandler(new ByteArrayOutputStream())
+      .option(
+        RuntimeOptions.LANGUAGE_HOME_OVERRIDE,
+        Paths.get("../../distribution/component").toFile().getAbsolutePath()
+      ).build();
+    final Map<String, Language> langs = ctx.getEngine().getLanguages();
+    assertNotNull("Enso found: " + langs, langs.get("enso"));
+
+    var code = """
```
Comment (radeusgd) [inline engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java:39]: Sorry that I'm writing post-mortem but I've only now figured out one more way to break this:

```
from Standard.Base import all
polyglot java import bar.Foo


type Ca_Type
    Ca x

    to_text self = Panic.throw "PANIK"

type Cd_Type
    Cd x

    to_text self = 23

type Cf_Type
    Cf x

    to_text self = self

main =
    z = Cd_Type.Cd 42
    Foo.bar z

    y = Ca_Type.Ca 42
    Foo.bar y

    x = Cf_Type.Cf 42
    Foo.bar x
```

on current develop (after this PR), yields:
```
[|Error in method `to_text` of [Cd 42]: Expected Text but got 23|]
[|Panic in method `to_text` of [Ca 42]: 'PANIK'|]
[|Error in toString()|]
```

I was testing this because I had a hunch that `to_text self = self` could lead to an infinite loop with the new logic. Apparently it doesn't, it just errors. For some reason though the error is different than what we handle so it still fallsback to the default Truffle `Error in toString()`.

I don't think it's a bad thing because it is a very rare case, I do not know how to repro this other than by returning `self`.

Still I thought it's nice to at least keep note of this fact.

Comment (JaroslavTulach) [reply engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java:39]: @radeusgd, please report the missing test case in pivotal as a bug.

Comment (radeusgd) [reply engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java:39]: OK

https://www.pivotaltracker.com/story/show/184324080
